src/main/java/com/linkedin/kmf/services/ProduceService.java

```diff
@@ -257,16 +288,18 @@ void addPartitionSensors(int partition) {
    */
   private class ProduceRunnable implements Runnable {
     private final int _partition;
+    private final String _key;
 
-    ProduceRunnable(int partition) {
+    ProduceRunnable(int partition, String key) {
       _partition = partition;
+      _key = key;
     }
 
     public void run() {
       try {
-        long nextIndex = _nextIndexPerPartition.get(_partition).get();
+        long nextIndex = _nextIndexPerKey.get(_key).get();
         String message = Utils.jsonFromFields(_topic, nextIndex, System.currentTimeMillis(), _producerId, _recordSize);
-        BaseProducerRecord record = new BaseProducerRecord(_topic, _partition, null, message);
+        BaseProducerRecord record = new BaseProducerRecord(_topic, null, _key, message);
```

Comment:
That should be fine. Partition number will override our key for as long as the partition is propagated through the pipeline but since they should correspond to each other, there shouldn't be issues